### PR TITLE
Sparse Table Implementation (Row, Col) -> Val

### PR DESCRIPTION
**Pull request by tomusdrw**

### CODE REVIEW

File: util/src/table.rs [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+		self.map.iter().fold(0, |acc, (_k, v)| acc + v.len())
+	}
+
+	/// Check if there is any element in this Table
+	pub fn is_empty(&self) -> bool {
+		self.len() == 0
+	}
+
+	/// Get mutable reference for single Table row.
+	pub fn get_row_mut(&mut self, row: &Row) -> Option<&mut HashMap<Col, Val>> {
```
Comment (arkpar) [inline util/src/table.rs]: Our convention is not to use `get_` prefix in function names. Should be just `row_mut`

File: util/src/table.rs [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+		self.map.get_mut(row)
+	}
+
+	/// Checks if row is defined for that table (note that even if defined it might be empty)
+	pub fn has_row(&self, row: &Row) -> bool {
+		self.map.contains_key(row)
+	}
+
+	/// Get immutable reference for single row in this Table
+	pub fn get_row(&self, row: &Row) -> Option<&HashMap<Col, Val>> {
```
Comment (arkpar) [inline util/src/table.rs]: `_get` prefix

File: util/src/table.rs [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+	}
+
+	/// Returns length of the Table (number of (row, col, val) tuples)
+	pub fn len(&self) -> usize {
+		self.map.iter().fold(0, |acc, (_k, v)| acc + v.len())
+	}
+
+	/// Check if there is any element in this Table
+	pub fn is_empty(&self) -> bool {
+		self.len() == 0
```
Comment (arkpar) [inline util/src/table.rs]: This is incorrect if there are empty rows

File: util/src/table.rs [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+		if is_empty {
+			self.map.remove(row);
+		}
+	}
+
+	/// Inserts new value to specified cell
+	///
+	/// Returns previous value (if any)
+	pub fn insert(&mut self, row: Row, col: Col, val: Val) -> Option<Val> {
+		if !self.map.contains_key(&row) {
```
Comment (arkpar) [inline util/src/table.rs]: Idiomatic way is to use `map.entry(row).or_insert_with(...)`

File: util/src/table.rs [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+	}
+
+	/// Removes all elements from this Table
+	pub fn clear(&mut self) {
+		self.map.clear();
+	}
+
+	/// Returns length of the Table (number of (row, col, val) tuples)
+	pub fn len(&self) -> usize {
+		self.map.iter().fold(0, |acc, (_k, v)| acc + v.len())
```
Comment (arkpar) [inline util/src/table.rs]: Clippy usually gives a warning to use 'values()` instead of `iter() ` in such cases

File: util/src/table.rs (hunk 3 above)
Comment (tomusdrw) [inline util/src/table.rs]: I think it's ok (note it's not `self.map.len()`)

Comment (arkpar) [inline util/src/table.rs]: Ah, true

File: util/src/table.rs (hunk 4 above)
Comment (tomusdrw) [inline util/src/table.rs]: Wow! Awesome! Thanks!

File: util/src/table.rs (hunk 3 above)
Comment (tomusdrw) [inline util/src/table.rs]: Changed the implementation to be more efficient (and more explicit)

**Pull request merged**